test/Npgsql.Tests/Replication/PgOutputReplicationTests.cs

```diff
@@ -0,0 +1,537 @@
+using Npgsql.Replication.Logical.Protocol;
+using NpgsqlTypes;
+using NUnit.Framework;
+
+namespace Npgsql.Tests.Replication
+{
+    public class PgOutputReplicationTests : SafeReplicationTestBase<NpgsqlLogicalReplicationConnection>
+    {
+        [Test]
+        public Task CreateReplicationSlot()
+            => SafeReplicationTest(nameof(CreateReplicationSlot) + "_pgoutput",
+                async (slotName, _) =>
+                {
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    await using var rc = await OpenReplicationConnectionAsync();
+                    var options = await rc.CreateReplicationSlot(slotName);
+
+                    using var cmd =
+                        new NpgsqlCommand($"SELECT * FROM pg_replication_slots WHERE slot_name = '{options.SlotName}'",
+                            c);
+                    await using var reader = await cmd.ExecuteReaderAsync();
+
+                    Assert.That(reader.Read, Is.True);
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("slot_type")), Is.EqualTo("logical"));
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("plugin")), Is.EqualTo("pgoutput"));
+                    Assert.That(reader.Read, Is.False);
+                });
+
+        [Test(Description = "Tests whether INSERT commands get replicated as Logical Replication Protocol Messages")]
+        public Task StartReplicationReplicatesInsert()
+            => SafeReplicationTest(nameof(StartReplicationReplicatesInsert) + "_pgoutput",
+                async (slotName, tableName, publicationName) =>
+                {
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    var messages = new ConcurrentQueue<LogicalReplicationProtocolMessage>();
+                    await c.ExecuteNonQueryAsync(@$"
+CREATE TABLE {tableName} (id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY, name TEXT NOT NULL);
+CREATE PUBLICATION {publicationName} FOR TABLE {tableName};
+");
+                    var rc = await OpenReplicationConnectionAsync();
+                    var slot = await rc.CreateReplicationSlot(slotName);
+                    await c.ExecuteNonQueryAsync($"INSERT INTO {tableName} (name) VALUES ('val1'), ('val2')");
+
+                    using var streamingCts = new CancellationTokenSource();
+                    var replicationTask = Task.Run(async () =>
+                    {
+                        await foreach (var msg in (await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token)).WithCancellation(streamingCts.Token))
```

Comment:
> Is there any reason we can't just process the results of StartReplication directly (i.e. await on the enumerator we get from its IAsyncEnumerable)?

We can, as long as we don't disable automated feedback (`WalReceiverStatusInterval = Timeout.InfiniteTimeSpan`).
If you do this the server may cut the stream because it requested feedback and didn't get a response in time.

In general we should warn people from sitting on an active replication stream without constantly polling it.
It's not only the feedback problem there's also the problem that the backend will not stop streaming and will fill up it's send buffers and disk space, depending on the load.

So this is rather meant as best practice guidance although the only test where it's really necessary is `SynchronousReplication()`.

Reply:
I understand that it's important to read frequently, but I really hope the best practice guidance isn't to set up a concurrent queue or channel (or any sort of buffer) in front of the replication IAsyncEnumerable... (if it really were, the right thing would have been to integrate this inside Npgsql implicitly).

Setting up a concurrent queue to avoid the server from filling up basically moves the problem from the server to the client - I'm not sure if that necessarily makes the problem better. Either your buffer is unlimited in size, in which case you'll simply run out of memory on the client, or it's limited in size, in which case you start to either drop messages (big no-no), or you stop reading messages into the buffer, and are back to the same original problem.

So this is a very classic problem of backpressure... It's indeed the client's responsibility to read at a reasonable rate - with reasonable being dependent on the load characteristics of the particular database being replicated. If the client *consistently* reads too slowly compared to the database update rate, then no buffer or concurrent queue in the world will help... And if the client is just *temporarily* not reading fast enough because of a load spike, or not reading at all, then I hope we can trust PG replication to properly buffer on its side. After all, the client application may be totally down for some time (reboot, maintenance, whatever), and as long as there's enough disk space on the PG side things should be fine, no?

Re automated feedback, I haven't yet looked at that part of the code - but if I'm understanding you correctly, by default Npgsql will send periodic pings (as feedback) even if the client doesn't read, right? If that's correct, is there any reason why a client would want to disable automated feedback?

Reply:
FYI https://github.com/Brar/npgsql/pull/4 removes the queue from the tests.